src/python/pants/backend/jvm/tasks/jvm_compile/compile_context.py

```diff
@@ -8,18 +8,23 @@
 import zipfile
 from contextlib import contextmanager
 
+from pants.backend.jvm.targets.jar_library import JarLibrary
 from pants.build_graph.aliased_target import AliasTarget
 from pants.build_graph.target import Target
 from pants.util.contextutil import open_zip
 
 
 def _resolve_strict_dependencies(target):
   for declared in target.dependencies:
-    if type(declared) in (AliasTarget, Target):
+    # The transitive dependencies of JarLibrary targets have
+    # to be included regardless of strict_deps setting,
+    # because they are always required for compilation.
+    if type(declared) in (AliasTarget, JarLibrary, Target):
```

Comment:
The funny thing about implementing it this way is that it changes the ordering of the resulting classpath entries.

I think a more intention revealing way to implement this would be to add an `exports` property to `JarLibrary` that's aliased to its dependencies.